### PR TITLE
Add Bayesian Optimisation to AB testing example

### PR DESCRIPTION
**Pull request by ae-foster**
To complete our first OED example, I've added a GP-based Bayesian optimizer based on http://pyro.ai/examples/bo.html . This uses Thompson sampling to acquire new data in batches.

Summary of changes:
 - `ab_test.py` - optimizing the APE using `GPBayesOptimizer`
 - `gp_bayes_opt.py` - new code implementing the BO example with Thompson sampling: probably need to move this somewhere else or can we merge it with existing BO code?
 - `gpr.py` - added a method to lazily create a multivariate sample from a Gaussian process posterior

### CODE REVIEW

**Comment by fritzo** [comment]
@ae-foster to get CI tests to run, you'll need to merge dev into your branch and then push.

File: examples/contrib/oed/gp_bayes_opt.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+        :param function differentiable: a function amenable to torch
+            autograd
+        :param torch.Tensor x_init: the initial point
+
+        """
+        # transform x to an unconstrained domain
+        unconstrained_x_init = transform_to(self.constraints).inv(x_init)
+        unconstrained_x = torch.tensor(unconstrained_x_init, requires_grad=True)
+        minimizer = optim.LBFGS([unconstrained_x], max_eval=20)
```
Comment (ae-foster) [inline examples/contrib/oed/gp_bayes_opt.py]: Cap evals at `20` to avoid some numerical instability

File: examples/contrib/oed/gp_bayes_opt.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+        def closure():
+            minimizer.zero_grad()
+            if (torch.log(torch.abs(unconstrained_x)) > 25.).any():
```
Comment (ae-foster) [inline examples/contrib/oed/gp_bayes_opt.py]: Somewhat arbitrary cutoff- necessary to prevent overflow during LBFGS run

File: examples/contrib/oed/gp_bayes_opt.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    #     with torch.no_grad():
+    #         loc, var = self.gpmodel(Xnew, full_cov=False, noiseless=False)
+    #         sd = var.sqrt()
+    #         ax1.plot(Xnew.numpy(), loc.numpy(), "r", lw=2)  # plot predictive mean
+    #         ax1.fill_between(Xnew.numpy(), loc.numpy() - 2*sd.numpy(), loc.numpy() + 2*sd.numpy(),
+    #                          color="C0", alpha=0.3)  # plot uncertainty intervals
+    #     ax1.set_xlim(-1, 101)
+    #     ax1.set_title("Find {}".format(xlabel))
+    #     if with_title:
+    #         ax1.set_ylabel("Gaussian Process Regression")
```
Comment (ae-foster) [inline examples/contrib/oed/gp_bayes_opt.py]: Commented code is unsightly but useful for debugging. `test_examples` will not admit `matplotlib` in example code.

File: pyro/contrib/gp/models/gpr.py [hunk 4]
```diff
@@ -141,3 +142,80 @@ def forward(self, Xnew, full_cov=False, noiseless=True):
...
+            Kffnew = Kff.new_empty(N+1, N+1)
+            Kffnew[:N, :N] = Kff
+            cross = self.kernel(X, xnew).squeeze()
+            end = self.kernel(xnew, xnew).squeeze()
+            Kffnew[N, :N] = cross
+            Kffnew[:N, N] = cross
+            # No noise, just jitter for numerical stability
+            Kffnew[N, N] = end + self.jitter
+            # Heuristic to avoid adding degenerate points
+            if Kffnew.logdet() > -15.:
```
Comment (ae-foster) [inline pyro/contrib/gp/models/gpr.py]: Arbitrary cutoff: prevent `Kff` becoming singular.

File: pyro/contrib/gp/models/gpr.py [hunk 5]
```diff
@@ -141,3 +142,80 @@ def forward(self, Xnew, full_cov=False, noiseless=True):
...
+            # Compute conditional mean and variance
+            loc, cov = conditional(xnew, X, self.kernel, y_residual, None, Lff,
+                                   False, jitter=self.jitter)
+            if not noiseless:
+                cov = cov + noise
+
+            # Todo use pyro.sample
+            d = normal.Normal(torch.tensor(0.), torch.tensor(1.))
+            # Reparametrize explicitly - aids autograd
+            ynew = (loc + self.mean_function(xnew)) + d.sample()*cov.sqrt()
```
Comment (ae-foster) [inline pyro/contrib/gp/models/gpr.py]: It would be nice to make this a true pyronic sampler if possible

File: examples/contrib/oed/gp_bayes_opt.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        """
+        self.f = f
+        self.constraints = constraints
+        self.gpmodel = gpmodel
+
+    def update_posterior(self, X, y):
+        X = torch.cat([self.gpmodel.X, X])
+        y = torch.cat([self.gpmodel.y, y])
+        self.gpmodel.set_data(X, y)
+        self.gpmodel.optimize()
```
Comment (ae-foster) [inline examples/contrib/oed/gp_bayes_opt.py]: Not very pyronic

File: examples/contrib/oed/gp_bayes_opt.py (hunk 3 above)
Comment (fritzo) [reply examples/contrib/oed/gp_bayes_opt.py]: Feel free to use local imports for plotting dependencies:
```py
def plot(self, gs, ...):
    from matplotlib import pyplot as plt
    ...
```
That's not perfect, but it's better than commented-out code :wink:

File: pyro/contrib/gp/models/gpr.py (hunk 5 above)
Comment (eb8680) [reply pyro/contrib/gp/models/gpr.py]: Why can't you write `torch.distributions.Normal(loc + self.mean_function(xnew), cov.sqrt()).rsample()`?

File: pyro/infer/trace_elbo.py [hunk 7]
```diff
@@ -115,7 +115,7 @@ def loss_and_grads(self, model, guide, *args, **kwargs):
 
             if trainable_params and getattr(surrogate_elbo_particle, 'requires_grad', False):
                 surrogate_loss_particle = -surrogate_elbo_particle / self.num_particles
-                surrogate_loss_particle.backward()
+                surrogate_loss_particle.backward(retain_graph=True)
```
Comment (eb8680) [inline pyro/infer/trace_elbo.py]: This change has some performance and memory implications.  I'll push a fix today as part of #1227 that exposes a `differentiable_loss` function that you can use when you need `retain_graph=True`

File: pyro/contrib/gp/models/gpr.py [hunk 8]
```diff
@@ -141,3 +142,80 @@ def forward(self, Xnew, full_cov=False, noiseless=True):
...
+            kernel's parameters have been learned from a training procedure (MCMC or
+            SVI).
+
+        :param bool noiseless: A flag to decide if we want to add sampling noise
+            to the samples beyond the noise inherent in the GP posterior.
+        :returns: sampler
+        :rtype: function
+        """
+        # Make these visible in the inner function
+        global X, y, Kff, N
```
Comment (eb8680) [inline pyro/contrib/gp/models/gpr.py]: Instead of using `global` here and below, how about making `sample_next` take these variables as arguments and returning a curried function with `functools.partial`?

```python
def sample_next(X, y, Kff, N, xnew):
    ...

return functools.partial(sample_next, X, y, Kff, xnew)
```

File: examples/contrib/oed/gp_bayes_opt.py (hunk 6 above)
Comment (eb8680) [reply examples/contrib/oed/gp_bayes_opt.py]: Can we do some GP magic to make these updates cheaper than recomputing the full posterior?

File: examples/contrib/oed/gp_bayes_opt.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+        :param str method: the method to use for acquisition. Choose from:
+            Thompson
+        :param int num_acquisitions: the number of points to generate
+        :param dict opt_params: additional parameters for optimization
+            routines
+        :return: a tensor of points to evaluate `self.f` at
+        :rtype: torch.Tensor
+        """
+
+        if method == "Thompson":
```
Comment (eb8680) [inline examples/contrib/oed/gp_bayes_opt.py]: Let's split this out into an acquisition function parameter instead of implementing acquisition functions inside `acquire`, so it's easier to experiment with different Monte Carlo acquisition functions.

File: examples/contrib/oed/gp_bayes_opt.py [hunk 10]
```diff
@@ -0,0 +1,164 @@
+# import matplotlib.gridspec as gridspec
+# import matplotlib.pyplot as plt
+import torch
+import torch.autograd as autograd
+import torch.optim as optim
+from torch.distributions import transform_to
+
+
+class GPBayesOptimizer:
```
Comment (eb8680) [inline examples/contrib/oed/gp_bayes_opt.py]: We should be able to refactor this a bit into a `pyro.optim.multi.MultiOptimizer` so that it looks more like other Pyro optimizers.  That way when we experiment with other optimizers e.g. gradient-based optimizers it'll be easy to swap them out

File: examples/contrib/oed/gp_bayes_opt.py (hunk 3 above)
Comment (eb8680) [reply examples/contrib/oed/gp_bayes_opt.py]: You should probably just add a separate test for this and remove the test code here

File: examples/contrib/oed/gp_bayes_opt.py [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+    """Performs Bayesian Optimization using a Gaussian Process as an
+    emulator for the unknown function.
```
Comment (eb8680) [inline examples/contrib/oed/gp_bayes_opt.py]: Once the code is a little more settled down, you should expand this docstring and add a couple usage examples

File: examples/contrib/oed/gp_bayes_opt.py [hunk 12]
```diff
@@ -0,0 +1,164 @@
+# import matplotlib.gridspec as gridspec
+# import matplotlib.pyplot as plt
```
Comment (eb8680) [inline examples/contrib/oed/gp_bayes_opt.py]: Remove these as discussed below

File: examples/contrib/oed/gp_bayes_opt.py (hunk 3 above)
Comment (ae-foster) [reply examples/contrib/oed/gp_bayes_opt.py]: I'll remove the code, keep it on a local branch

File: pyro/contrib/gp/models/gpr.py (hunk 5 above)
Comment (ae-foster) [reply pyro/contrib/gp/models/gpr.py]: Yes, had it in my mind that that didn't work- it does

File: examples/contrib/oed/gp_bayes_opt.py (hunk 6 above)
Comment (ae-foster) [reply examples/contrib/oed/gp_bayes_opt.py]: Yes depends how much magic you want. Here https://github.com/uber/pyro/blob/dev/pyro/contrib/gp/models/gpr.py#L80 we could have cached `Kff` and just add the kernel computations we need. One step further is to use the magic of Schur complements https://en.wikipedia.org/wiki/Schur_complement to avoid having to invert the whole kernel matrix. I actually started implementing a Schur complement method for `iter_sample` but in the end the existing solution works and reuses more existing code. In either case, we would have to refactor internals of `GPRegression`

File: pyro/infer/trace_elbo.py (hunk 7 above)
Comment (ae-foster) [reply pyro/infer/trace_elbo.py]: Kk I'll rebase from `dev` once that is merged

File: pyro/contrib/gp/models/gpr.py (hunk 8 above)
Comment (ae-foster) [reply pyro/contrib/gp/models/gpr.py]: The reason I didn't do that was that we change `X` inside the inner function (by writing `X = Xnew`). I thought scoping rules would mean that that change wasn't saved, but honestly I didn't actually try it

File: examples/contrib/oed/gp_bayes_opt.py (hunk 6 above)
Comment (eb8680) [reply examples/contrib/oed/gp_bayes_opt.py]: Let's open a separate issue about this, it seems like generally useful functionality for `contrib.gp`

File: pyro/infer/trace_elbo.py (hunk 7 above)
Comment (eb8680) [reply pyro/infer/trace_elbo.py]: Added in #1252

File: examples/contrib/oed/gp_bayes_opt.py (hunk 1 above)
Comment (ae-foster) [reply examples/contrib/oed/gp_bayes_opt.py]: Sure :) I didn't realize that was in the pipeline

File: examples/contrib/oed/gp_bayes_opt.py (hunk 10 above)
Comment (ae-foster) [reply examples/contrib/oed/gp_bayes_opt.py]: I think we're best to talk over this together- we need to implement a `step` function which would replace my `run` function. Would be also need to refactor the example to use `pyro.param`? Right now, it's not really true that after a `step`, the params are updated to new near-optimal values. But obviously I could add `self.opt_differentiable(lambda x: self.gpmodel(x)[0])` into each step (minimize the current GP mean function)

File: examples/contrib/oed/gp_bayes_opt.py (hunk 6 above)
Comment (ae-foster) [reply examples/contrib/oed/gp_bayes_opt.py]: https://github.com/uber/pyro/issues/1258

File: pyro/infer/trace_elbo.py (hunk 7 above)
Comment (eb8680) [reply pyro/infer/trace_elbo.py]: Let's remove this change and switch to using `TraceEnum_ELBO` below

File: pyro/contrib/gp/models/gpr.py (hunk 8 above)
Comment (eb8680) [reply pyro/contrib/gp/models/gpr.py]: You can also put these variables into a dictionary and mutate that instead of using `global`

**Comment by ae-foster** [comment]
@eb8680 have the CI passing now :)
